core/server/worker/src/main/java/alluxio/worker/block/TieredBlockStore.java

```diff
@@ -526,42 +697,92 @@ private void abortBlockInternal(long sessionId, long blockId) throws BlockDoesNo
-      BlockStoreLocation loc;
-      String srcPath;
-      String dstPath;
-      TempBlockMeta tempBlockMeta;
-      try (LockResource r = new LockResource(mMetadataReadLock)) {
-        checkTempBlockOwnedBySession(sessionId, blockId);
-        tempBlockMeta = mMetaManager.getTempBlockMeta(blockId);
-        srcPath = tempBlockMeta.getPath();
-        dstPath = tempBlockMeta.getCommitPath();
-        loc = tempBlockMeta.getBlockLocation();
+      // Allocate from given location.
+      dirView = mAllocator.allocateBlockWithView(sessionId, options.getSize(),
+          options.getLocation(), allocatorView);
+
+      if (dirView != null) {
+        return dirView;
       }
 
-      // Heavy IO is guarded by block lock but not metadata lock. This may throw IOException.
-      FileUtils.move(srcPath, dstPath);
+      if (options.isForceLocation()) {
+        if (options.isEvictionAllowed()) {
+          freeSpace(sessionId, options.getSize(), options.getSize(), options.getLocation());
+          dirView = mAllocator.allocateBlockWithView(sessionId, options.getSize(),
+              options.getLocation(), allocatorView.refreshView());
 
-      try (LockResource r = new LockResource(mMetadataWriteLock)) {
-        mMetaManager.commitTempBlockMeta(tempBlockMeta);
-      } catch (BlockAlreadyExistsException | BlockDoesNotExistException
-          | WorkerOutOfSpaceException e) {
-        throw Throwables.propagate(e); // we shall never reach here
-      }
+          if (dirView == null) {
+            LOG.error("Target tier: {} has no evictable space to store {} bytes for session: {}",
+                options.getLocation(), options.getSize(), sessionId);
+            return null;
+          }
+        } else {
+          LOG.error("Target tier: {} has no available space to store {} bytes for session: {}",
+              options.getLocation(), options.getSize(), sessionId);
+          return null;
+        }
+      } else {
+        dirView = mAllocator.allocateBlockWithView(sessionId, options.getSize(),
+            BlockStoreLocation.anyTier(), allocatorView);
 
-      // Check if block is pinned on commit
-      if (pinOnCreate) {
-        addToPinnedInodes(BlockId.getFileId(blockId));
-      }
+        if (dirView != null) {
+          return dirView;
+        }
 
-      return loc;
-    } finally {
-      mLockManager.unlockBlock(lockId);
+        if (options.isEvictionAllowed()) {
+          // There is no space left on worker.
+          // Free more than requested by configured free-ahead size.
+          long freeAheadBytes =
+              ServerConfiguration.getBytes(PropertyKey.WORKER_TIERED_STORE_FREE_AHEAD_BYTES);
+          freeSpace(sessionId, options.getSize(), options.getSize() + freeAheadBytes,
+              BlockStoreLocation.anyTier());
+
+          dirView = mAllocator.allocateBlockWithView(sessionId, options.getSize(),
+              BlockStoreLocation.anyTier(), allocatorView.refreshView());
```

Comment:
Should we do a check if dirView is null here and do our failure handling (ie. log message now but I suggest to increment some metric)?

Reply:
Currently, as per previous trend, callers all throw/log for null response from this method. I think metrics would be fine. However I was planning an all over work on defining new metrics. It can be done at the end of this PR or as a separate PR.